packages/frontend/component/src/components/attachment-viewer/viewer.tsx

```diff
@@ -193,62 +159,74 @@ export const Viewer = ({ model }: ViewerProps): ReactElement => {
 
     setDocInfo(info => {
-      const cursor = Math.ceil(scrollTop / (scrollHeight / info.total));
-      // thumbnailsScrollerHandleRef.current?.scrollToIndex(cursor)
-      return {
-        ...info,
-        cursor,
-      };
+      const itemHeight = scrollHeight / info.total;
+      const n = scrollTop / itemHeight;
+      const t = n / info.total;
+      const index = Math.floor(n + t);
+      const cursor = Math.min(index, info.total - 1);
+
+      if (cursor === info.cursor) return info;
+      return { ...info, cursor };
     });
-    // }, [scrollerRef, thumbnailsScrollerHandleRef]);
   }, [scrollerRef]);
 
   const onSelect = useCallback(
     (index: number) => {
-      scrollerHandleRef.current?.scrollToIndex(index);
-      setDocInfo(info => ({ ...info, cursor: index }));
+      scrollerHandleRef.current?.scrollToIndex({
+        index,
+        align: 'start',
+        behavior: 'smooth',
+      });
     },
     [scrollerHandleRef]
   );
 
   const updateMainVisibleRange = useMemo(
-    () => debounce(setMainVisibleRange, 233, { leading: true, trailing: true }),
+    () => debounce(setMainVisibleRange, 233, { trailing: true }),
     [setMainVisibleRange]
   );
 
   const updateThumbnailsVisibleRange = useMemo(
-    () =>
-      debounce(setThumbnailsVisibleRange, 233, {
-        leading: true,
-        trailing: true,
-      }),
+    () => debounce(setThumbnailsVisibleRange, 233, { trailing: true }),
     [setThumbnailsVisibleRange]
   );
 
-  // useEffect(() => {
-  //   const el = viewerRef.current;
-  //   if (!el) return;
+  useEffect(() => {
+    const el = viewerRef.current;
+    if (!el) return;
+
+    return observeResize(el, entry => {
+      const rect = entry.contentRect;
+      setViewportInfo(info => ({
+        ...info,
+        width: rect.width,
+        height: rect.height,
+      }));
+    });
+  }, [viewerRef]);
 
-  //   return observeResize(el, entry => {
-  //     console.log(entry);
-  //   });
-  // }, []);
+  useEffect(() => {
+    post(MessageOp.SyncViewportInfo, viewportInfo);
+  }, [viewportInfo, post]);
 
   useEffect(() => {
     const { startIndex, endIndex } = mainVisibleRange;
     let index = startIndex;
-    for (; index < endIndex + 1; index++) {
-      post(MessageOp.Render, { index, kind: 'page' });
+    for (; index <= endIndex; index++) {
+      post(MessageOp.Render, { index, kind: RenderKind.Page });
     }
   }, [mainVisibleRange, post]);
 
   useEffect(() => {
+    if (collapsed) return;
+
     const { startIndex, endIndex } = thumbnailsVisibleRange;
+
     let index = startIndex;
-    for (; index < endIndex + 1; index++) {
-      post(MessageOp.Render, { index, kind: 'thumbnail' });
+    for (; index <= endIndex; index++) {
+      post(MessageOp.Render, { index, kind: RenderKind.Thumbnail });
     }
-  }, [thumbnailsVisibleRange, post]);
+  }, [collapsed, thumbnailsVisibleRange, post]);
 
   useEffect(() => {
     workerRef.current = new Worker(
```

Comment:
rewrite the logic here into an util class. `WorkerController`

Reply:
this will be processed together with the blocksuite pdf service extension